### ISSUE TITLE
[COURSE]: Differentiable Programming

### ISSUE THREAD

**Issue by jatkinson1000** [Cambridge-ICCS/summer-school-guidance#10]
### Course Name

Differentiable Programming

### Description

Derivatives are at the heart of scientific programming. From the Jacobian matrices used to solve nonlinear systems to the gradient vectors used for optimisation methods, from the backpropagation operation in machine learning to the data
assimilation methods used in weather forecasting, all of these techniques rely on derivative information. Differentiable programming (also known as automatic/algorithmic differentiation (AD)) provides a suite of tools for users to compute derivatives of quantities in their code without any manual encoding. In Session 1, we will learn about the history and mathematical background of differentiable programming and consider examples using the [Autograd](https://github.com/HIPS/autograd) AD tool. In Session 2, we will learn about more advanced topics and consider examples using the [JAX](https://docs.jax.dev/en/latest/) differentiable modelling framework.

### Pre-requisites

* Undergraduate level knowledge of linear algebra and calculus.
* Basic knowledge of Python.
* A GitHub account.

### Course Materials Repository/Link

https://github.com/Cambridge-ICCS/differentiable-programming-summer-school-2026

### Leader(s)

@joewallwork 

### Contributors

@sjavis
@AdelekeBankole 

### Progress Checklist Session 1

- [x] Repository/materials created
- [x] Content checked against guidance
- [x] Practice session run
- [ ] Feedback incorporated

### Practice Session 1
- **Date:** 21/May/2026 (10:00-12:00)
- **Participants:** @isaacaka 

#### Practice Session 1 Feedback

* https://github.com/Cambridge-ICCS/differentiable-programming-summer-school-2026/issues/32
* https://github.com/Cambridge-ICCS/differentiable-programming-summer-school-2026/issues/33
* https://github.com/Cambridge-ICCS/differentiable-programming-summer-school-2026/issues/34
* https://github.com/Cambridge-ICCS/differentiable-programming-summer-school-2026/issues/36

### Progress Checklist Session 2

- [ ] Repository/materials created
- [ ] Content checked against guidance
- [ ] Practice session run
- [ ] Feedback incorporated

### Practice Session 2
- **Date:** 29/June/2026 (14:00-16:00)
- **Participants:** @isaacaka

**Comment by dorchard** [comment]
I believe the run throughs schedule for:
Session 1 with examples in Autograd - 10:00-12:00 on Thursday 21st May
Session 2 with examples in JAX - 13:00-15:00 on Monday 29th June
is that right @joewallwork ?

**Comment by joewallwork** [comment]
Yes, that's right @dorchard

**Comment by jatkinson1000** [comment]
Notes from runthrough:

## jwa34 notes from runthrough

Running locally on Mac with Python 3.13

### General Notes

- Too much is packed in here
  - Distill down to key takeaways and allow learning/reinforcment of these.
- Dark mode in notebook is harder to read in runthroughs.
  Perhaps switch to light for workshop.
- You often say "I'll leave you to convince yourselves that's correct" or
  "I'll leave that as an optional exercise to check" and
  immediately move on. Either pause for people to do so, or remove?
  Reduces the feeling of learning and makes it appear that we are behind.
- Lots of "You can X but I won't go into what that is".
  Carpentries would say don't mention it al all then!
  similar to avoiding "basically", "just", etc.
- In general it feels too fast, and that you are explaining to someone also
  well-versed in DP rather than someone new.
- Feels like most of the session was just clicking through cells.
  ~1h before I really wrote any code.
- What was the aim of this session!?
  - To introduce JAX?
    - Maybe... grad an JVP functions
  - Learn about forward and reverse modes applied to functions of several variables.
    - Yes
  - See showcases of more advanced AD usage.
    - ??
  - Learn about checkpointing and using AD to compute higher order derivatives.
    - Yes, but no exercise. Drop this (see comment below)? 

### Specific points

- Local install I get:
  ```
  /Users/jwa34/differential-prog/differentiable-programming-summer-school-2026/.venv/lib/python3.13/site-packages/mpltools/annotation/_slopemarker.py:12: SyntaxWarning: invalid escape sequence '\|'
  """Plot triangular slope marker labeled with slope.
  ```
  for cell 2, though it goes away if I re-run the cell.
- Get rid of "derivatived in code" text
- Clarify "WHY" arrays are immutable?
  To avoid shooting yourself in the foot/ensure operation chain remains
  intact.
- JAX help - the note that you can make a new functioin from the grad of a
  function was rushed over quite quickly and didn't make sense to me.
  Perhaps don't mention (as we do this anyway) or let people read in
  their own time?
- Move higher order derivatives later in the notebook
- For seeding, I think adding a code block to demonstrate what you are
  talking about would help a lot, as it comes very thing and fast when just
  talking. And would break up the front-heavy talking.
- f and g optional example I think it would be useful to spend a couple of
  mins letting people think through this. Alternatively, I'm not sure what
  this illustrates.
- Chaining forward mode - keep note that we can write as product of three
  JVP, but remove the discussion of additional notation.
  - Or... can we use examples of just 2 operations to make it easier to follow?
- "In my mind vjp is something different but let's not go there"
  Perhaps don't mention at all unless people are likely to get confused
  elsewhere in future?
- Talk about f with the DAG, but f is not on there. Perhaps place a reminder
  on the image of what f and g are etc?
- Computing the Hessian, some sort of diagram showing the two routes (forward/reverse) and
  where they go from/to with what data would be nice.
  - Add a TeX representation of the Hessian to aid you when explaining?
- the explanation ov vmap was confusing to me. I kept trying to
  vmap(grad)(tanh(x)) instead of vmap(grad(tanh))(x)
  Perhaps a basic "chaining vmap" example before this would help clarify use?
- Do we need the section on source transformation? I'm not sure it adds to
  the session and just adds more confusion? Especially when you did a full
  explanation of the code and talked about static analysis and the various
  options out there etc.
  Could you just note that "as an alternative we could have written the code
  for derivatives alongside (like Data Assimilation) but JAX has already done
  this"?
- NN example
  - exercise 1 add a docstring of what the input x is as a hint?
  - Wasn't clear to me exercise 2 didn't expect code.
  - exercise 3 it is set up to expect non-vmap.
    - It wasn't clear to me that we were going to use blank image as input.
      Only clear oncve I saw the output (which is really cool).
- exercise 4/5 went very fast, wasn't sure what the difference between the two
  approaches was.
- I think checkpointing is just a step too much.
  - People who try and use this in a project will find this out when they need it?
